### ISSUE TITLE
Calculate kW out of kWh

### ISSUE THREAD

**Issue by Sineos** [graphite-project/graphite-web#1112]
Hi all,

I'm using graphite to plot some energy meters (power consumption). Now I have a meter that gives the total consumed power, e.g. 100 kWh. I query this meter every 20 seconds so the total is constantly increasing when power is consumed.
I would like to calculate the current wattage in graphite and have it displayed as graph. Somehow I cannot wrap my head around this and all my noobish attempts with summarize and derivative ended in crappy results. Any help would be really appreciated.

Kind regards,

Sineos

**Comment by obfuscurity** [comment]
So I'm an idiot when it comes to electricity. But it sounds like you have a counter value for wattage. If you want to view the current *rate*, just use the `derivative()` or `nonNegativeDerivative()` functions around that metric. If that doesn't get you what you're expecting to see, please elaborate on your results for each attempt.

**Comment by Sineos** [comment]
Many thanks for this first reply. But I could not get plausible values from the derivative nor the summarize functions or a combination here of.
Maybe an example of the mathematical model helps:

Lets assume an electrical device that 2000 Watt power.
You run the device for 2 hours.
Now you have consumed 4 kWh.
During this 2 hours my metric would increase from its [current value] to [current value + 4] in small steps.
The graph I intend to get should read 2000 during this, which is the wattage of the electrical device.

**Comment by brutasse** [comment]
It's probably a resolution issue. What's the time resolution of your data? You're probably seeing 2000 / (number of datapoints in an hour).

**Comment by Sineos** [comment]
I capture the data currently every 20 seconds. Lets assume that my meter has currently a value of 100 kWh. Given the above example, this total would increase by 0.005555556 kWh every 20 seconds and would reach 104 after 7200 seconds (= 2h).

**Comment by brutasse** [comment]
And 20 seconds is also the resolution of your whisper files?

What happens if you take the result of your derivative and do `scale(xxx, 180)`? (180 being the number of datapoints you have in a 1h)

**Comment by Sineos** [comment]
Yes, 20 seconds matches the whisper file

When analysing it, the "derivative(home.power.ehz.in2)" is at least close to the expected result with respect to the shape of the graph. I cannot give it more precisely but the result is about a factor of 180000 too low. This might somehow be related to the 180 measuring points and the difference between me expecting Watts and the meter counting kilo Watts hours. --> 180 * 1000 = 180000??

scale(home.power.ehz.in2, 180) is way off

but ....

scale(derivative(home.power.ehz.in2), 180000) does the trick and with pretty high accuracy but some small aberrations that remain a mystery
![clipboard01](https://cloud.githubusercontent.com/assets/4459312/5805708/a35aaf6a-a011-11e4-9ea4-6f99c3149d54.png)

This image is from a test meter that actually delivers both the value (Power in Watts) I'm searching and the total consumption in kWh.

Many thanks for pointing me into the right direction :+1:

**Comment by brutasse** [comment]
Looks good! Not sure why you're seeing these aberrations though.

**Comment by Sineos** [comment]
Thanks a lot for this awesome tool and the support.
The features and possibilities are still quite overwhelming to me and honestly the above results is based on trial and error and not on deeper understanding of the tool but nonetheless I'm totally enthusiastic with it.

**Comment by Sineos** [comment]
My endeavors on electrical measurement continue. Currently I'm using a test meter that displays more values than later my actual setup will to verify the results and test for plausibility of the values.

integral(scale(home.power.ehz.power, 0.000005555555555))
This function gives me the opposite way of what I have done before with the derivative function. It takes the current power usage in Watt and provides the total consumption over time. The 5.555e-6 value in this case is the reciprocal of the 180,000 used above. So far so good, this calculation works and gives me exact results.

Now I wanted to generate a daily consumption as total. So lets assume at the start of the day the graphs last value was 100, at the end of the day it was 119. This means I used 19 kWh and the value I want to get from the graphite functions would exactly be this 19. I used:

summarize(derivative(integral(scale(home.power.ehz.power, 0.000005555555555))), '24h', 'sum', false)

or for better readability:

summarize(derivative(calculation_above), '24h', 'sum', false)

I expected to get 19 as result, but it turned out to be 16. Anyone an idea where this big deviation is coming from?
